libbpf-tools/biostacks.c

```diff
@@ -0,0 +1,213 @@
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.duration = strtoll(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "invalid delay (in us): %s\n", arg);
+			argp_usage(state);
+		}
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static int print_map(struct ksyms *ksyms, int fd)
+{
+	char *units = env.milliseconds ? "msecs" : "usecs";
+	struct rqinfo lookup_key = {}, next_key;
+	const struct ksym *ksym;
+	int num_stack, i, err;
+	struct hist hist;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		printf("%-14.14s %-6d %-7s\n",
+			next_key.comm, next_key.pid, next_key.disk);
+		num_stack = next_key.kern_stack_size /
+			sizeof(next_key.kern_stack[0]);
+		for (i = 0; i < num_stack; i++) {
+			ksym = ksyms__map_addr(ksyms, next_key.kern_stack[i]);
+			printf("%s\n", ksym->name);
+		}
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		printf("\n");
+		lookup_key = next_key;
+	}
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct ksyms *ksyms = NULL;
+	struct biostacks_bpf *obj;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = biostacks_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF ojbect\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	if (env.disk)
+		strncpy((char*)obj->rodata->targ_disk, env.disk, env.disk_len);
+	obj->rodata->targ_ms = env.milliseconds;
+
+	err = biostacks_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	obj->links.fentry__blk_account_io_start =
+		bpf_program__attach(obj->progs.fentry__blk_account_io_start);
+	err = libbpf_get_error(obj->links.fentry__blk_account_io_start);
+	if (err) {
+		fprintf(stderr, "failed to attach blk_account_io_start: %s\n",
+			strerror(err));
+		goto cleanup;
+	}
+	ksyms = ksyms__load();
+	if (!ksyms) {
+		fprintf(stderr, "failed to load kallsyms\n");
+		goto cleanup;
+	}
+	if (ksyms__get_symbol(ksyms, "blk_account_io_merge_bio")) {
+		obj->links.kprobe__blk_account_io_merge_bio =
+			bpf_program__attach(obj->
+					progs.kprobe__blk_account_io_merge_bio);
+		err = libbpf_get_error(obj->
+				links.kprobe__blk_account_io_merge_bio);
```

Comment:
I really think you should drop "kprobe__" and "fentry__" prefixes, will make for shorter identifiers and they are just redundant and not-so-important prefixes, IMO

Reply:
Sure, I'll drop these prefixes, and drop other tools in next PR. Thanks!